### PR TITLE
MF-15 funding secured page

### PR DESCRIPTION
**Pull request by scottywakefield**
notable omission - doesn't show any details of the created reservation on the confirmation page.

### CODE REVIEW

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationCommandHandler.cs [hunk 1]
```diff
@@ -1,24 +1,58 @@
...
+                throw new ArgumentException(
+                    "The following parameters have failed validation", 
+                    validationResult.ValidationDictionary
+                        .Select(c => c.Key)
+                        .Aggregate((item1, item2) => item1 + ", " + item2));
+            }
 
-            return new CreateReservationResult();
+            var decodedAccountId = _hashingService.DecodeValue(command.AccountId);
+            var apiRequest = new CreateReservationApiRequest
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationCommandHandler.cs]: Can we change this to be something like:
```
var apiRequest = new CreateReservationApiRequest(_hashingService.DecodeValue,command.AccountId,command.StartDate);
```
which in turn will mean that the model will look like:
```
public class CreateReservationApiRequest
    {
        public CreateReservationApiRequest(Func<string,long> setAccountId, string employerAccountId, DateTime startDate)
        {
            AccountId = setAccountId(employerAccountId);
            StartDate = startDate;
        }
        public long AccountId { get; }
        public DateTime StartDate { get;}
    }
```

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationApiRequest.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace SFA.DAS.Reservations.Application.Reservations.Commands
+{
+    public class CreateReservationApiRequest
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationApiRequest.cs]: Can this be moved - would be better in the domain

Also if we could change it to be like:
```
public class CreateReservationApiRequest
    {
        public CreateReservationApiRequest(Func<string,long> setAccountId, string employerAccountId, DateTime startDate)
        {
            AccountId = setAccountId(employerAccountId);
            StartDate = startDate;
        }
        public long AccountId { get; }
        public DateTime StartDate { get;}
    }
```
Where the setAccountId is the _hashingService.Decode

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationCommandHandler.cs [hunk 3]
```diff
@@ -1,24 +1,58 @@
...
+            {
+                AccountId = decodedAccountId,
+                StartDate = command.StartDate
+            };
+
+            var reservationJson = await _apiClient.CreateReservation(decodedAccountId, JsonConvert.SerializeObject(apiRequest));
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CreateReservationCommandHandler.cs]: Can the API client be changed to take a generic type where we can set the Reservation?

File: src/SFA.DAS.Reservations.Infrastructure/Api/ApiClient.cs [hunk 4]
```diff
@@ -16,14 +17,30 @@ public ApiClient(IOptions<ReservationsApiConfiguration> apiOptions)
...
+                response.EnsureSuccessStatusCode();
+
+                return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            }
+        }
+
+        public async Task<string> CreateReservation(long accountId, string json)
+        {
+            var accessToken = await GetAccessTokenAsync();
+            using (var client = new HttpClient())//not unit testable using directly
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Infrastructure/Api/ApiClient.cs]: Can we whack an interface around this? Or pull in the SFA.DAS.SecureHttp package?

File: src/SFA.DAS.Reservations.Web.UnitTests/Reservations/WhenCallingPostCreate.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+using SFA.DAS.Reservations.Application.Reservations.Commands;
+using SFA.DAS.Reservations.Web.Controllers;
+
+namespace SFA.DAS.Reservations.Web.UnitTests.Reservations
+{
+    [TestFixture]
+    public class WhenCallingPostCreate
+    {
+        [Test]
+        public async Task Then_Sends_Create_Reservation_Command()
```
[the remark this answers is not in the document]
Comment (scottywakefield) [reply src/SFA.DAS.Reservations.Web.UnitTests/Reservations/WhenCallingPostCreate.cs]: done

File: src/SFA.DAS.Reservations.Web/Controllers/ReservationsController.cs [hunk 6]
```diff
@@ -8,10 +12,39 @@ namespace SFA.DAS.Reservations.Web.Controllers
...
+        public async Task<IActionResult> Create()
+        {
+            var accountId = RouteData.Values["employerAccountId"].ToString();
+            var command = new CreateReservationCommand
+            {
+                AccountId = accountId,
+                StartDate = DateTime.Today
+            };
+
+            var createResult = await _mediator.Send(command);
```
[the remark this answers is not in the document]
Comment (scottywakefield) [reply src/SFA.DAS.Reservations.Web/Controllers/ReservationsController.cs]: done

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/WhenCreatingANewReservation.cs [hunk 7]
```diff
@@ -1,45 +1,118 @@
...
+                .ReturnsAsync(new ValidationResult());
+
+            _mockApiClient = fixture.Freeze<Mock<IApiClient>>();
+            _mockApiClient
+                .Setup(client => client.CreateReservation(It.IsAny<long>(), It.IsAny<string>()))
+                .ReturnsAsync(reservationJson);
+
+            _mockHashingService = fixture.Freeze<Mock<IHashingService>>();
+            _mockHashingService
+                .Setup(service => service.DecodeValue(It.IsAny<string>()))
```
[the remark this answers is not in the document]
Comment (scottywakefield) [reply src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/WhenCreatingANewReservation.cs]: I'm using a different value for a lot of the tests, but the specific functionality is tested in `Then_Decodes_The_AccountId`.

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/WhenValidatingACreateReservationCommand.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                StartDate = DateTime.Today
+            };
+
+            var result = await validator.ValidateAsync(command);
+
+            result.IsValid().Should().BeFalse();
+            result.ValidationDictionary.Count.Should().Be(1);
+            result.ValidationDictionary
+                .Should().ContainKey(nameof(CreateReservationCommand.AccountId))
+                .WhichValue.Equals($"{nameof(CreateReservationCommand.AccountId)} has not been supplied");
```
[the remark this answers is not in the document]
Comment (scottywakefield) [reply src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Commands/WhenValidatingACreateReservationCommand.cs]: `.WhichValue` refers to the value of the `.ContainsKey()` so it goes one step further and enforces that the key and value pair are set up correctly.

File: src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Services/WhenEncodingAndDecoding.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            var fixture = new Fixture()
+                .Customize(new AutoMoqCustomization{ConfigureMembers = true});
+
+            _decodedId = fixture.Create<long[]>();
+            _encodedId = fixture.Create<string>();
+            
+            _mockHashIds = fixture.Freeze<Mock<IHashids>>();
+            _mockHashIds
+                .Setup(hashids => hashids.DecodeLong(It.IsAny<string>()))
```
[the remark this answers is not in the document]
Comment (scottywakefield) [reply src/SFA.DAS.Reservations.Application.UnitTests/Reservations/Services/WhenEncodingAndDecoding.cs]: done

**Pull request merged**